cpp/src/arrow/csv/writer.cc

```diff
@@ -108,6 +108,17 @@ int64_t CountQuotes(std::string_view s) {
 constexpr int64_t kQuoteCount = 2;
 constexpr int64_t kQuoteDelimiterCount = kQuoteCount + /*end_char*/ 1;
 
+#ifndef CSV_WRITER_DISPATCH_STRING_ARRAY_TYPE
+#define CSV_WRITER_DISPATCH_STRING_ARRAY_TYPE(array, func, ...) \
+  do {                                                          \
+    if (ARROW_PREDICT_TRUE(array->type_id() == Type::STRING)) { \
+      return func<StringArray>(__VA_ARGS__);                    \
+    } else {                                                    \
+      return func<LargeStringArray>(__VA_ARGS__);               \
```

Comment:
Also dcheck the type is large_string here?

Reply:
Fixed.